pyiron_contrib/tinybase/container.py

```diff
@@ -0,0 +1,179 @@
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+class StorageAttribute:
+    """
+    Create an attribute that is synced to a storage attribute.
+
+    It must be created on a class that derives from :class:`.HasStorage`.  It essentially behaves like a property that
+    writes and reads values to the underlying :attr:`.HasStorage.storage`. DataContainer of the class.  When accessing
+    this property before setting it, `None` is returned.
+
+    It's possible to modify the default value and the accepted type of values by using the builder-style :meth:`.type`
+    and :meth:`.default` methods.
+
+    >>> class MyType(HasStorage):
+    ...     a = StorageAttribute()
+    ...     b = StorageAttribute().type(int)
+    ...     c = StorageAttribute().default(list)
+    ...     d = StorageAttribute().default(lambda: 42).type(int)
+
+    >>> t = MyType()
+    >>> t.a # return None
+    >>> t.b = 3
+    >>> t.b
+    3
+    >>> t.b = 'asdf'
+    TypeError(f"{value} is not of type {self.value_type}")
+    >>> t.c
+    []
+    >>> t.d
+    42
+    """
+
+    def __init__(self):
+        self.name = None
+        self.value_type = object
+        self.default_constructor = None
+
+    def __set_name__(self, obj, name):
+        self.name = name
+
+    def __get__(self, obj, objtype=None):
+        try:
+            return obj.storage[self.name]
+        except KeyError:
+            if self.default_constructor is not None:
+                ret = obj.storage[self.name] = self.default_constructor()
+                return ret
+            else:
+                return None
+
+    def __set__(self, obj, value):
+        if isinstance(value, self.value_type):
+            obj.storage[self.name] = value
+        else:
+            raise TypeError(f"{value} is not of type {self.value_type}")
+
+    def type(self, value_type):
+        """
+        Set a type to type check values set on the attribute.
+
+        Args:
+            value_type (type, tuple of type): a class or list of classes that are acceptable for the attribute
+
+        Returns:
+            self: the object it is called on
+        """
+        self.value_type = value_type
+        return self
+
+    def default(self, default_constructor):
+        """
+        Set a function to create a default value, if the attribute is accessed without being set.
+
+        Args:
+            default_constructor (function, type): Either a class or a function that takes no arguments
+
+        Returns:
+            self: the object it is called on
+        """
+        self.default_constructor = default_constructor
+        return self
+
+    def doc(self, text):
+        self.__doc__ = text
+
+class AbstractContainer(HasStorage, HasHDFAdapaterMixin, abc.ABC):
+    # TODO: this should go into HasStorage, exists here only to give one location to define from_ methods
+    @classmethod
+    def from_attributes(cls, name, *attrs, module=None, bases=(), **default_attrs):
```

Comment:
Dynamic class creation makes me a little uncomfortable. Is there no more vanilla way to accomplish what you're looking for? My guess is that this is here for the syntactic sugar of making new input classes very quickly, but if we can find something, say, <33% longer but avoiding this magic I would take that hands-down.

Reply:
Sure,

```python
class MyInput(AbstractContainer):
    field1 = StorageAttribute()
    field2 = StorageAttribute().type(int).default(42)
```
is the same as the corresponding `from_attributes` call.  I just got lazy prototyping.  I was thinking it'd be cute to be able to make an input class from a `dict` of default and this is where `from_attributes` comes from, but I didn't follow through with it.

Reply:
I think the crux of the issue here is that I'm somewhat torn between the ability to do stuff my composition when we explicitly define IO classes, and the elegance/stability of generating IO automatically from the actual behaviour of the node.

I think there is a direct analogy here where I _used_ to (and, e.g. ryven3 does) have templates for node IO for each of the individual IO variables. You could package these templates together and build them by composition, in exact analogy to what you do here. Just like here, you wind up needing to explicitly state type variables, defaults, etc. Then your trick with `from_attributes` becomes helpful syntactic sugar for making this all easier to type.

What @samwaseda really pushed me to over in the graph PR is instead to _automatically_ scrape as much of this (arg name, type(optional), default(optional)) from the functional definition of the node behaviour itself. (Note: the definition is functional, but if you pass around mutable data as arguments you still have the freedom/danger of non-functional graphs.)

Maybe there is a way to think about getting some sort of function composition going to have the best of both worlds, because after he pushed me to go that route, I would up reeeaaally liking Sam's information-extraction approach.

Reply:
I think there's a good middle.  Actually `from_attributes` would work nicely together with getting names and types from the function definition, so you could generate these on the fly quite easily.  What you lose  I think is then the ability to compose it nicely and safely.  E.g. a Murnaghan needs to know that it can just stick a structure into a node and get an energy out.   In a notebook I agree its very convenient to derive it from a annotations, but for more long lived code I'd prefer a bit more OOP.  I do think we can have both by making the compatibility check between take into account first the class relations and then whether matching attributes are defined like you do in the graphs.